File: lab4/main.py
class RegexParserError(Exception):
    pass

class Token:
    def __init__(self, ttype, value=None):
        self.ttype = ttype
        self.value = value

    def __repr__(self):
        return f"Token({self.ttype}, {self.value})"

class Lexer:
    def __init__(self, text):
        self.text = text
        self.pos = 0

    def peek(self):
        if self.pos < len(self.text):
            return self.text[self.pos]
        return None

    def advance(self):
        self.pos += 1

    def tokenize(self):
        tokens = []
        while self.pos < len(self.text):
            ch = self.peek()

            if ch == '(':
                # Проверяем конструкции (?:  (?=  (?N)
                self.advance()
                nxt = self.peek()
                if nxt == '?':
                    # Это либо незахватывающая, либо опережающая проверка, либо (?N)
                    self.advance()
                    nxt2 = self.peek()
                    if nxt2 == ':':
                        # (?: )
                        self.advance()
                        tokens.append(Token('NONCAP_OPEN'))
                    elif nxt2 == '=':
                        # (?= )
                        self.advance()
                        tokens.append(Token('LOOKAHEAD_OPEN'))
                    elif nxt2 and nxt2.isdigit():
                        # (?N)
                        self.advance()  # перешли на цифру N
                        val = int(nxt2)
                        tokens.append(Token('EXPR_REF_OPEN', val))
                        # Закрывающая скобка будет обработана в парсере
                    else:
                        raise RegexParserError("Некорректный синтаксис после (?")
                else:
                    # Захватывающая группа ( ... )
                    tokens.append(Token('CAP_OPEN'))
            elif ch == ')':
                tokens.append(Token('CLOSE'))
                self.advance()
            elif ch == '|':
                tokens.append(Token('ALT'))
                self.advance()
            elif ch == '*':
                tokens.append(Token('STAR'))
                self.advance()
            elif ch and 'a' <= ch <= 'z':
                tokens.append(Token('CHAR', ch))
                self.advance()
            else:
                raise RegexParserError(f"Неизвестный символ: {ch}")
        return tokens


# Узлы AST
class GroupNode:
    def __init__(self, group_id, node):
        self.group_id = group_id
        self.node = node

    def __repr__(self):
        return f"GroupNode({self.group_id}, {self.node})"

class NonCapGroupNode:
    def __init__(self, node):
        self.node = node

    def __repr__(self):
        return f"NonCapGroupNode({self.node})"

class LookaheadNode:
    def __init__(self, node):
        self.node = node

    def __repr__(self):
        return f"LookaheadNode({self.node})"

class ConcatNode:
    def __init__(self, nodes):
        self.nodes = nodes

    def __repr__(self):
        return f"ConcatNode({self.nodes})"

class AltNode:
    def __init__(self, branches):
        self.branches = branches

    def __repr__(self):
        return f"AltNode({self.branches})"

class StarNode:
    def __init__(self, node):
        self.node = node

    def __repr__(self):
        return f"StarNode({self.node})"

class CharNode:
    def __init__(self, ch):
        self.ch = ch

    def __repr__(self):
        return f"CharNode('{self.ch}')"

class ExprRefNode:
    def __init__(self, ref_id):
        self.ref_id = ref_id

    def __repr__(self):
        return f"ExprRefNode({self.ref_id})"

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0
        self.group_count = 0
        self.max_groups = 9
        self.in_lookahead = False

        # Сохраняем определения групп, чтобы проверять корректность ссылок
        # group_id -> AST подграмматики
        self.groups_ast = {}

    def current_token(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def eat(self, ttype=None):
        tok = self.current_token()
        if tok is None:
            raise RegexParserError("Неожиданный конец выражения")
        if ttype is not None and tok.ttype != ttype:
            raise RegexParserError(f"Ожидается {ttype}, найдено {tok.ttype}")
        self.pos += 1
        return tok

    def parse(self):
        node = self.parse_alternation()
        if self.current_token() is not None:
            # Если что-то осталось непрочитанное, синтаксическая ошибка
            raise RegexParserError("Лишние символы после корректного выражения")
        # Проверяем синтаксическую корректность, но не будем запрещать forward refs
        self.check_references(node, defined_groups=set())
        return node

    def parse_alternation(self):
        # alternation: concatenation ('|' concatenation)*
        branches = [self.parse_concatenation()]
        while self.current_token() and self.current_token().ttype == 'ALT':
            self.eat('ALT')
            if self.current_token() is None or self.current_token().ttype in ['CLOSE', 'ALT']:
                raise RegexParserError("Пустая альтернатива запрещена")
            branches.append(self.parse_concatenation())
        if len(branches) == 1:
            return branches[0]
        return AltNode(branches)

    def parse_concatenation(self):
        # concatenation: repetition+
        nodes = []
        while self.current_token() and self.current_token().ttype not in ['CLOSE', 'ALT']:
            nodes.append(self.parse_repetition())
        if len(nodes) == 1:
            return nodes[0]
        return ConcatNode(nodes)

    def parse_repetition(self):
        # repetition: base ('*')?
        node = self.parse_base()
        while self.current_token() and self.current_token().ttype == 'STAR':
            self.eat('STAR')
            node = StarNode(node)
        return node

    def parse_base(self):
        tok = self.current_token()
        if tok is None:
            raise RegexParserError("Неожиданный конец при ожидании базового выражения")

        if tok.ttype == 'CAP_OPEN':
            # ( ... )
            self.eat('CAP_OPEN')
            self.group_count += 1
            if self.group_count > self.max_groups:
                raise RegexParserError("Превышено число групп захвата > 9")
            group_id = self.group_count
            node = self.parse_alternation()
            self.eat('CLOSE')
            # Сохраняем AST группы
            self.groups_ast[group_id] = node
            return GroupNode(group_id, node)

        elif tok.ttype == 'NONCAP_OPEN':
            # (?: ... )
            self.eat('NONCAP_OPEN')
            node = self.parse_alternation()
            self.eat('CLOSE')
            return NonCapGroupNode(node)

        elif tok.ttype == 'LOOKAHEAD_OPEN':
            # (?= ... )
            if self.in_lookahead:
                raise RegexParserError("Вложенные опережающие проверки запрещены")
            self.eat('LOOKAHEAD_OPEN')
            old_look = self.in_lookahead
            self.in_lookahead = True
            node = self.parse_alternation()
            self.in_lookahead = old_look
            self.eat('CLOSE')
            return LookaheadNode(node)

        elif tok.ttype == 'EXPR_REF_OPEN':
            # (?N)
            ref_id = tok.value
            self.eat('EXPR_REF_OPEN')
            self.eat('CLOSE')
            return ExprRefNode(ref_id)

        elif tok.ttype == 'CHAR':
            ch = tok.value
            self.eat('CHAR')
            return CharNode(ch)

        else:
            raise RegexParserError(f"Некорректный токен: {tok}")

    def check_references(self, node, defined_groups):
        # Проверка корректности:
        # 1) ExprRefNode(ref_id): ref_id должна быть определена (ref_id in defined_groups)
        # 2) GroupNode: после неё добавить её номер
        # 3) LookaheadNode: внутри не может быть групп захвата, других lookahead
        # PS:
        # Не запрещаем forward references, рекурсия допустима
        # Просто проверим отсутствия lookahead внутри lookahead или групп захвата в нём.
        if isinstance(node, CharNode):
            return defined_groups

        elif isinstance(node, ExprRefNode):
            # Не выдаём ошибку при forward ref
            return defined_groups

        elif isinstance(node, GroupNode):
            # Внутри группы сначала проверяем содержимое
            new_defined = self.check_references(node.node, defined_groups)
            # После конца группы эта группа считается определённой
            new_defined = set(new_defined)
            new_defined.add(node.group_id)
            return new_defined

        elif isinstance(node, NonCapGroupNode):
            return self.check_references(node.node, defined_groups)

        elif isinstance(node, LookaheadNode):
            # Проверить, что внутри нет групп захвата и нет других lookahead
            self.check_no_cap_and_lookahead(node.node, inside_lookahead=True)
            # Ссылки на группы должны быть из уже определённых
            return self.check_references(node.node, defined_groups)

        elif isinstance(node, StarNode):
            return self.check_references(node.node, defined_groups)

        elif isinstance(node, ConcatNode):
            cur_defined = defined_groups
            for child in node.nodes:
                cur_defined = self.check_references(child, cur_defined)
            return cur_defined

        elif isinstance(node, AltNode):
            # Изначально было пересечение, теперь делаем объединение,
            # чтобы ситуации вроде (a|(bb))(a|(?2)) были корректными.
            all_defs = []
            for branch in node.branches:
                branch_defs = self.check_references(branch, defined_groups)
                all_defs.append(branch_defs)
            union_defs = set()
            for d in all_defs:
                union_defs.update(d)
            return union_defs

        else:
            raise RegexParserError("Неизвестный тип узла AST при проверке ссылок")

    def check_no_cap_and_lookahead(self, node, inside_lookahead):
        """
        Проверяем, что внутри lookahead нет захватывающих групп и lookahead.
        """
        if isinstance(node, GroupNode) and inside_lookahead:
            raise RegexParserError("Внутри опережающей проверки не допускаются захватывающие группы")
        if isinstance(node, LookaheadNode) and inside_lookahead:
            raise RegexParserError("Внутри опережающей проверки не допускаются другие опережающие проверки")

        if isinstance(node, (NonCapGroupNode, LookaheadNode, StarNode, ConcatNode, AltNode)):
            # Рекурсивно проверяем для детей
            if isinstance(node, NonCapGroupNode):
                self.check_no_cap_and_lookahead(node.node, inside_lookahead)
            elif isinstance(node, LookaheadNode):
                self.check_no_cap_and_lookahead(node.node, inside_lookahead)
            elif isinstance(node, StarNode):
                self.check_no_cap_and_lookahead(node.node, inside_lookahead)
            elif isinstance(node, ConcatNode):
                for n in node.nodes:
                    self.check_no_cap_and_lookahead(n, inside_lookahead)
            elif isinstance(node, AltNode):
                for b in node.branches:
                    self.check_no_cap_and_lookahead(b, inside_lookahead)

# ------------------------------------------------------------
# Построение каркасной КС-грамматики
# ------------------------------------------------------------

class CFGBuilder:
    def __init__(self, groups_ast):
        # groups_ast: {group_id: node}
        self.groups_ast = groups_ast
        self.group_nonterm = {}
        self.noncap_index = 1
        self.star_index = 1

    def build(self, node):
        start = 'S'
        rules = {}

        # Регистрируем G1 для группы 1
        main_nt = self.node_to_cfg(node, rules)

        # Добавляем правило для S -> main_nt (группа 1)
        rules[start] = [[main_nt]]

        # Убедимся, что все группы зарегистрированы
        for group_id, ast in self.groups_ast.items():
            if group_id not in self.group_nonterm:
                nt = f"G{group_id}"
                self.group_nonterm[group_id] = nt
                self.node_to_cfg(ast, rules, start_symbol=nt)

        return start, rules

    def node_to_cfg(self, node, rules, start_symbol=None):
        """
        Преобразует узел AST в нетерминал CFG.
        Возвращает имя нетерминала.
        rules: dict {NonTerminal: [ [symbols], [symbols] ... ]}
        """
        if isinstance(node, CharNode):
            # Терминальный символ
            # Если нам дали start_symbol, используем его как имя нетерминала, иначе генерим
            nt = start_symbol if start_symbol else self.fresh_nt('CHAR')
            rules.setdefault(nt, []).append([node.ch])
            return nt

        elif isinstance(node, GroupNode):
            nt = self.group_nonterm.get(node.group_id)
            if nt is None:
                nt = f"G{node.group_id}"
                self.group_nonterm[node.group_id] = nt
            # Строим внутреннее правило для содержимого группы
            sub_nt = self.node_to_cfg(node.node, rules)
            # G{group_id} -> sub_nt
            rules.setdefault(nt, []).append([sub_nt])
            return nt

        elif isinstance(node, NonCapGroupNode):
            # Генерируем новый нетерминал для незахватывающей группы
            nt = start_symbol if start_symbol else self.fresh_nt('N')
            sub_nt = self.node_to_cfg(node.node, rules)
            # Правило: nt -> sub_nt
            rules.setdefault(nt, []).append([sub_nt])
            return nt

        elif isinstance(node, LookaheadNode):
            # Опережающую проверку не представляем в CFG (по условию), заменим ее на ε
            nt = start_symbol if start_symbol else self.fresh_nt('LA')
            rules.setdefault(nt, []).append([])
            return nt

        elif isinstance(node, ConcatNode):
            nt = start_symbol if start_symbol else self.fresh_nt('C')
            # Конкатенация: nodes - список узлов
            # Преобразуем каждый узел в нетерминал, потом nt -> seq
            # Если узел терминальный, node_to_cfg вернёт нетерминал с одним правилом
            seq_nts = [self.node_to_cfg(ch, rules) for ch in node.nodes]
            rules.setdefault(nt, []).append(seq_nts)
            return nt

        elif isinstance(node, AltNode):
            nt = start_symbol if start_symbol else self.fresh_nt('A')
            # Альтернатива: для каждой ветви генерируем правило
            for branch in node.branches:
                br_nt = self.node_to_cfg(branch, rules)
                # nt -> br_nt
                rules.setdefault(nt, []).append([br_nt])
            return nt

        elif isinstance(node, StarNode):
            # Звёздочка: X* означает 0 или более повторений X
            # Создаём нетерминал для звёздочки
            nt = start_symbol if start_symbol else self.fresh_nt('R')
            sub_nt = self.node_to_cfg(node.node, rules)
            # R -> ε | R sub_nt
            rules.setdefault(nt, []).append([])          # ε
            rules[nt].append([nt, sub_nt])
            return nt

        elif isinstance(node, ExprRefNode):
            # Ссылка на выражение группы
            ref_id = node.ref_id
            if ref_id not in self.group_nonterm:
                self.group_nonterm[ref_id] = f"G{ref_id}"
                if ref_id not in self.groups_ast:
                    raise RegexParserError(f"Ссылка на несуществующую группу {ref_id}")
                # Строим правила для группы ref_id
                sub_nt = self.node_to_cfg(self.groups_ast[ref_id], rules)
                nt = self.group_nonterm[ref_id]
                # sub_nt уже построен выше:
                rules.setdefault(nt, []).append([sub_nt])
            return self.group_nonterm[ref_id]

        else:
            raise RegexParserError("Неизвестный тип узла при построении грамматики")

    def fresh_nt(self, prefix):
        # Генерируем уникальный нетерминал
        # Можно использовать простые счётчики
        if prefix == 'N':
            name = f"N{self.noncap_index}"
            self.noncap_index += 1
            return name
        elif prefix == 'R':
            name = f"R{self.star_index}"
            self.star_index += 1
            return name
        else:
            # Для простоты: добавим общий счётчик
            name = f"{prefix}{self.noncap_index + self.star_index}"
            self.noncap_index += 1
            return name


def main():
    text = input().strip()
    try:
        if text == "":
            raise RegexParserError("Пустая строка")
        # Разбиваем на лексеммы
        lexer = Lexer(text)
        tokens = lexer.tokenize()

        # Строим дерево
        parser = Parser(tokens)
        ast = parser.parse()

        # Построим КС-грамматику
        builder = CFGBuilder(parser.groups_ast)
        start_symbol, rules = builder.build(ast)
        print("Выражение корректно синтаксически и удовлетворяет ограничениям.")
        print("Построенная КС-грамматика (каркас):")
        print("Начальный нетерминал:", start_symbol)
        for nt in rules:
            for rhs in rules[nt]:
                rhs_str = " ".join(rhs) if rhs else "ε"
                print(f"{nt} -> {rhs_str}")
        print("=" * 60)
        for token in tokens:
            print(token)
        print("=" * 60)
        print(ast)

    except RegexParserError as e:
        print("Ошибка:", e)

if __name__ == "__main__":
    main()

File: lab4/main.jl
# Исключение для ошибок
struct RegexParserError <: Exception
    msg::String
end

# Тип для представления токена
struct Token
    ttype::String
    value::Union{Nothing, Int, String}
end

function Base.show(io::IO, t::Token)
    print(io, "Token(", t.ttype, ", ", t.value, ")")
end

# Лексер
mutable struct Lexer
    text::String
    pos::Int
end

# Конструктор
function Lexer(text::String)
    return Lexer(text, 1)
end

# Получить текущий символ
function peek(lexer::Lexer)
    return lexer.pos <= length(lexer.text) ? lexer.text[lexer.pos] : nothing
end

# Продвинуть указатель
function advance!(lexer::Lexer)
    lexer.pos += 1
end

# Основной метод для токенизации
function tokenize(lexer::Lexer)
    tokens = Token[]
    while lexer.pos <= length(lexer.text)
        ch = peek(lexer)

        if ch == '('
            advance!(lexer)
            nxt = peek(lexer)
            if nxt == '?'
                advance!(lexer)
                nxt2 = peek(lexer)
                if nxt2 == ':'
                    advance!(lexer)
                    push!(tokens, Token("NONCAP_OPEN", nothing))
                elseif nxt2 == '='
                    advance!(lexer)
                    push!(tokens, Token("LOOKAHEAD_OPEN", nothing))
                elseif isnumeric(nxt2)
                    advance!(lexer)
                    val = Base.parse(Int, string(nxt2))
                    push!(tokens, Token("EXPR_REF_OPEN", val))
                else
                    throw(RegexParserError("Некорректный синтаксис после (?"))
                end
            else
                push!(tokens, Token("CAP_OPEN", nothing))
            end
        elseif ch == ')'
            push!(tokens, Token("CLOSE", nothing))
            advance!(lexer)
        elseif ch == '|'
            push!(tokens, Token("ALT", nothing))
            advance!(lexer)
        elseif ch == '*'
            push!(tokens, Token("STAR", nothing))
            advance!(lexer)
        elseif ch >= 'a' && ch <= 'z'
            push!(tokens, Token("CHAR", string(ch)))
            advance!(lexer)
        else
            throw(RegexParserError("Неизвестный символ: $ch"))
        end
    end
    return tokens
end

# Узлы AST

# Узел группы (захватывающей)
struct GroupNode
    group_id::Int
    node
end

function Base.show(io::IO, node::GroupNode)
    print(io, "GroupNode(", node.group_id, ", ", node.node, ")")
end

# Узел незахватывающей группы
struct NonCapGroupNode
    node
end

function Base.show(io::IO, node::NonCapGroupNode)
    print(io, "NonCapGroupNode(", node.node, ")")
end

# Узел опережающей проверки
struct LookaheadNode
    node
end

function Base.show(io::IO, node::LookaheadNode)
    print(io, "LookaheadNode(", node.node, ")")
end

# Узел конкатенации
struct ConcatNode
    nodes::Vector
end

function Base.show(io::IO, node::ConcatNode)
    print(io, "ConcatNode(", node.nodes, ")")
end

# Узел альтернативы
struct AltNode
    branches::Vector
end

function Base.show(io::IO, node::AltNode)
    print(io, "AltNode(", node.branches, ")")
end

# Узел повторения (*)
struct StarNode
    node
end

function Base.show(io::IO, node::StarNode)
    print(io, "StarNode(", node.node, ")")
end

# Узел символа
struct CharNode
    ch::String
end

function Base.show(io::IO, node::CharNode)
    print(io, "CharNode('", node.ch, "')")
end

# Узел ссылки на выражение (?N)
struct ExprRefNode
    ref_id::Int
end

function Base.show(io::IO, node::ExprRefNode)
    print(io, "ExprRefNode(", node.ref_id, ")")
end

# ------------------------------------------------
# Парсер
# ------------------------------------------------
mutable struct Parser
    tokens::Vector{Token}
    pos::Int
    group_count::Int
    max_groups::Int
    in_lookahead::Bool
    groups_ast::Dict{Int, Any}  # Определение групп: group_id -> AST
end

# Конструктор
function Parser(tokens::Vector{Token})
    return Parser(tokens, 1, 0, 9, false, Dict())
end

function current_token(parser::Parser)
    parser.pos <= length(parser.tokens) ? parser.tokens[parser.pos] : nothing
end

function eat!(parser::Parser, ttype::String = nothing)
    tok = current_token(parser)
    if isnothing(tok)
        throw(RegexParserError("Неожиданный конец выражения"))
    end
    if !isnothing(ttype) && tok.ttype != ttype
        throw(RegexParserError("Ожидается $ttype, найдено $(tok.ttype)"))
    end
    parser.pos += 1
    return tok
end

# Основной метод парсинга
function parse_ast(parser::Parser)
    node = parse_alternation(parser)
    if !isnothing(current_token(parser))
        throw(RegexParserError("Лишние символы после корректного выражения"))
    end
    return node
end

function parse_alternation(parser::Parser)
    branches = Any[parse_concatenation(parser)]
    while !isnothing(current_token(parser)) && current_token(parser).ttype == "ALT"
        eat!(parser, "ALT")
        push!(branches, parse_concatenation(parser))
    end
    return length(branches) == 1 ? branches[1] : AltNode(branches)
end

function parse_concatenation(parser::Parser)
    nodes = Any[]
    while !isnothing(current_token(parser)) && !(current_token(parser).ttype in ["CLOSE", "ALT"])
        push!(nodes, parse_repetition(parser))
    end
    return length(nodes) == 1 ? nodes[1] : ConcatNode(nodes)
end


function parse_repetition(parser::Parser)
    node = parse_base(parser)
    while !isnothing(current_token(parser)) && current_token(parser).ttype == "STAR"
        eat!(parser, "STAR")
        node = StarNode(node)
    end
    return node
end

function parse_base(parser::Parser)
    tok = current_token(parser)
    if isnothing(tok)
        throw(RegexParserError("Неожиданный конец при ожидании базового выражения"))
    end

    if tok.ttype == "CAP_OPEN"
        eat!(parser, "CAP_OPEN")
        parser.group_count += 1
        if parser.group_count > parser.max_groups
            throw(RegexParserError("Превышено число групп захвата > 9"))
        end
        group_id = parser.group_count
        node = parse_alternation(parser)
        eat!(parser, "CLOSE")
        parser.groups_ast[group_id] = node
        return GroupNode(group_id, node)

    elseif tok.ttype == "NONCAP_OPEN"
        eat!(parser, "NONCAP_OPEN")
        node = parse_alternation(parser)
        eat!(parser, "CLOSE")
        return NonCapGroupNode(node)

    elseif tok.ttype == "LOOKAHEAD_OPEN"
        if parser.in_lookahead
            throw(RegexParserError("Вложенные опережающие проверки запрещены"))
        end
        eat!(parser, "LOOKAHEAD_OPEN")
        old_lookahead = parser.in_lookahead
        parser.in_lookahead = true

        # Вызов проверки на захватывающие группы внутри lookahead
        node = parse_alternation(parser)
        check_no_cap_and_lookahead(parser, node, true)

        parser.in_lookahead = old_lookahead
        eat!(parser, "CLOSE")
        return LookaheadNode(node)

    elseif tok.ttype == "EXPR_REF_OPEN"
        ref_id = tok.value
        eat!(parser, "EXPR_REF_OPEN")
        eat!(parser, "CLOSE")
        return ExprRefNode(ref_id)

    elseif tok.ttype == "CHAR"
        eat!(parser, "CHAR")
        return CharNode(tok.value)

    else
        throw(RegexParserError("Некорректный токен: $tok"))
    end
end


# Проверка корректности ссылок на группы и вложенности
function check_references(parser::Parser, node, defined_groups::Set{Int})
    if node isa CharNode
        return defined_groups

    elseif node isa ExprRefNode
        # Не проверяем forward references (допустимы)
        return defined_groups

    elseif node isa GroupNode
        # Проверяем содержимое группы
        new_defined = check_references(parser, node.node, defined_groups)
        # Добавляем текущую группу в множество определённых групп
        push!(new_defined, node.group_id)
        return new_defined

    elseif node isa NonCapGroupNode
        return check_references(parser, node.node, defined_groups)

    elseif node isa LookaheadNode
        # Проверяем, что внутри lookahead нет групп захвата или других lookahead
        check_no_cap_and_lookahead(parser, node.node, true)
        return check_references(parser, node.node, defined_groups)

    elseif node isa StarNode
        return check_references(parser, node.node, defined_groups)

    elseif node isa ConcatNode
        cur_defined = defined_groups
        for child in node.nodes
            cur_defined = check_references(parser, child, cur_defined)
        end
        return cur_defined

    elseif node isa AltNode
        all_defs = []
        for branch in node.branches
            push!(all_defs, check_references(parser, branch, defined_groups))
        end
        # Объединяем все определения групп из ветвей альтернативы
        union_defs = Set{Int}()
        for d in all_defs
            union!(union_defs, d)
        end
        return union_defs

    else
        throw(RegexParserError("Неизвестный тип узла AST при проверке ссылок"))
    end
end

# Проверка отсутствия групп захвата и вложенных lookahead внутри lookahead
function check_no_cap_and_lookahead(parser::Parser, node, inside_lookahead::Bool)
    if node isa GroupNode && inside_lookahead
        throw(RegexParserError("Внутри опережающей проверки не допускаются захватывающие группы"))
    elseif node isa LookaheadNode && inside_lookahead
        throw(RegexParserError("Внутри опережающей проверки не допускаются другие опережающие проверки"))
    end

    if node isa NonCapGroupNode
        check_no_cap_and_lookahead(parser, node.node, inside_lookahead)
    elseif node isa LookaheadNode
        check_no_cap_and_lookahead(parser, node.node, inside_lookahead)
    elseif node isa StarNode
        check_no_cap_and_lookahead(parser, node.node, inside_lookahead)
    elseif node isa ConcatNode
        for n in node.nodes
            check_no_cap_and_lookahead(parser, n, inside_lookahead)
        end
    elseif node isa AltNode
        for b in node.branches
            check_no_cap_and_lookahead(parser, b, inside_lookahead)
        end
    end
end


# Построитель КС-грамматики (CFG)
mutable struct CFGBuilder
    groups_ast::Dict{Int, Any}        # {group_id: AST узел}
    group_nonterm::Dict{Int, String}
    noncap_index::Int
    star_index::Int
end

# Конструктор
function CFGBuilder(groups_ast::Dict{Int, Any})
    return CFGBuilder(groups_ast, Dict{Int, String}(), 1, 1)
end

# Построение грамматики
function build(builder::CFGBuilder, node)
    start = "S"              # Начальный нетерминал
    rules = Dict{String, Vector{Vector{String}}}()  # Правила CFG

    # Регистрируем основной нетерминал для группы 1
    main_nt = node_to_cfg(builder, node, rules)

    # Добавляем правило S -> main_nt
    rules[start] = [[main_nt]]

    # Убедимся, что все группы зарегистрированы
    for (group_id, ast) in builder.groups_ast
        if group_id ∉ keys(builder.group_nonterm)
            nt = "G$group_id"
            builder.group_nonterm[group_id] = nt
            node_to_cfg(builder, ast, rules, start_symbol=nt)
        end
    end

    return start, rules
end

# Преобразование AST в нетерминал CFG
function node_to_cfg(builder::CFGBuilder, node, rules::Dict{String, Vector{Vector{String}}};
                     start_symbol::Union{String, Nothing} = nothing)
    if node isa CharNode
        # Терминальный символ
        nt = isnothing(start_symbol) ? fresh_nt(builder, "CHAR") : start_symbol
        push!(get!(rules, nt, []), [node.ch])
        return nt

    elseif node isa GroupNode
        # Захватывающая группа
        nt = get(builder.group_nonterm, node.group_id, nothing)
        if isnothing(nt)
            nt = "G$(node.group_id)"
            builder.group_nonterm[node.group_id] = nt
        end
        sub_nt = node_to_cfg(builder, node.node, rules)
        push!(get!(rules, nt, []), [sub_nt])
        return nt

    elseif node isa NonCapGroupNode
        # Незахватывающая группа
        nt = isnothing(start_symbol) ? fresh_nt(builder, "N") : start_symbol
        sub_nt = node_to_cfg(builder, node.node, rules)
        push!(get!(rules, nt, []), [sub_nt])
        return nt

    elseif node isa LookaheadNode
        # Lookahead заменяем на ε
        nt = isnothing(start_symbol) ? fresh_nt(builder, "LA") : start_symbol
        push!(get!(rules, nt, []), [])
        return nt

    elseif node isa ConcatNode
        # Конкатенация узлов
        nt = isnothing(start_symbol) ? fresh_nt(builder, "C") : start_symbol
        seq_nts = [node_to_cfg(builder, child, rules) for child in node.nodes]
        push!(get!(rules, nt, []), seq_nts)
        return nt

    elseif node isa AltNode
        # Альтернатива (ветви)
        nt = isnothing(start_symbol) ? fresh_nt(builder, "A") : start_symbol
        for branch in node.branches
            br_nt = node_to_cfg(builder, branch, rules)
            push!(get!(rules, nt, []), [br_nt])
        end
        return nt

    elseif node isa StarNode
        # Повторение (X*)
        nt = isnothing(start_symbol) ? fresh_nt(builder, "R") : start_symbol
        sub_nt = node_to_cfg(builder, node.node, rules)
        push!(get!(rules, nt, []), [])              # ε
        push!(rules[nt], [nt, sub_nt])              # R -> R sub_nt
        return nt

    elseif node isa ExprRefNode
        # Ссылка на выражение группы
        ref_id = node.ref_id
        if ref_id ∉ keys(builder.group_nonterm)
            builder.group_nonterm[ref_id] = "G$ref_id"
            if ref_id ∉ keys(builder.groups_ast)
                throw(RegexParserError("Ссылка на несуществующую группу $ref_id"))
            end
            node_to_cfg(builder, builder.groups_ast[ref_id], rules)
        end
        return builder.group_nonterm[ref_id]

    else
        throw(RegexParserError("Неизвестный тип узла при построении грамматики"))
    end
end

# Генерация имени нетерминала
function fresh_nt(builder::CFGBuilder, prefix::String)
    if prefix == "N"
        name = "N$(builder.noncap_index)"
        builder.noncap_index += 1
        return name
    elseif prefix == "R"
        name = "R$(builder.star_index)"
        builder.star_index += 1
        return name
    else
        # Общий счётчик для других типов
        name = "$(prefix)$(builder.noncap_index + builder.star_index)"
        builder.noncap_index += 1
        return name
    end
end

function main()
    text = String(strip(readline()))

    try
        if isempty(text)
            throw(RegexParserError("Пустая строка"))
        end

        # Разбиваем на лексемы
        lexer = Lexer(text)
        tokens = tokenize(lexer)

        # Строим AST
        parser = Parser(tokens)
        ast = parse_ast(parser)

        # Строим КС-грамматику
        builder = CFGBuilder(parser.groups_ast)
        start_symbol, rules = build(builder, ast)

        println("Выражение корректно синтаксически и удовлетворяет ограничениям.")
        println("Построенная КС-грамматика (каркас):")
        println("Начальный нетерминал: ", start_symbol)

        # Печатаем правила грамматики
        for (nt, rhs_list) in rules
            for rhs in rhs_list
                rhs_str = isempty(rhs) ? "ε" : join(rhs, " ")
                println("$nt -> $rhs_str")
            end
        end

        println("=" ^ 20) 
        for token in tokens
            println(token)
        end

        println("=" ^ 20)
        println(ast)

    catch e
        if e isa RegexParserError
            println("Ошибка: ", e.msg)
        else
            rethrow(e) 
        end
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

